pinot-broker/src/main/java/org/apache/pinot/broker/requesthandler/BrokerRequestHandlerDelegate.java

```diff
@@ -0,0 +1,113 @@
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.pinot.broker.requesthandler;
+
+import com.fasterxml.jackson.databind.JsonNode;
+import javax.annotation.Nullable;
+import org.apache.pinot.broker.api.RequesterIdentity;
+import org.apache.pinot.common.response.BrokerResponse;
+import org.apache.pinot.spi.trace.RequestContext;
+import org.apache.pinot.spi.utils.CommonConstants;
+
+
+/**
+ * {@code BrokerRequestHandlerDelegate} delegates the inbound broker request to one of the enabled
+ * {@link BrokerRequestHandler} based on the requested handle type.
+ *
+ * {@see: @CommonConstant
+ */
+public class BrokerRequestHandlerDelegate implements BrokerRequestHandler {
+
+  private String _defaultBrokerRequestHandlerType;
+  @Nullable
+  private SingleConnectionBrokerRequestHandler _singleConnectionBrokerRequestHandler;
+  @Nullable
+  private GrpcBrokerRequestHandler _grpcBrokerRequestHandler;
+  @Nullable
+  private WorkerQueryRequestHandler _multiStageWorkerRequestHandler;
+
+  public BrokerRequestHandlerDelegate(@Nullable String defaultBrokerRequestHandlerType,
+      @Nullable SingleConnectionBrokerRequestHandler singleConnectionBrokerRequestHandler,
+      @Nullable GrpcBrokerRequestHandler grpcBrokerRequestHandler,
+      @Nullable WorkerQueryRequestHandler multiStageWorkerRequestHandler
+  ) {
+    _defaultBrokerRequestHandlerType = defaultBrokerRequestHandlerType != null ? defaultBrokerRequestHandlerType
+        : CommonConstants.Broker.DEFAULT_BROKER_REQUEST_HANDLER_TYPE;
+    _singleConnectionBrokerRequestHandler = singleConnectionBrokerRequestHandler;
+    _grpcBrokerRequestHandler = grpcBrokerRequestHandler;
+    _multiStageWorkerRequestHandler = multiStageWorkerRequestHandler;
+  }
+
+  @Override
+  public void start() {
+    if (_singleConnectionBrokerRequestHandler != null) {
+      _singleConnectionBrokerRequestHandler.start();
+    }
+    if (_grpcBrokerRequestHandler != null) {
+      _grpcBrokerRequestHandler.start();
+    }
+    if (_multiStageWorkerRequestHandler != null) {
+      _multiStageWorkerRequestHandler.start();
+    }
+  }
+
+  @Override
+  public void shutDown() {
+    if (_singleConnectionBrokerRequestHandler != null) {
+      _singleConnectionBrokerRequestHandler.shutDown();
+    }
+    if (_grpcBrokerRequestHandler != null) {
+      _grpcBrokerRequestHandler.shutDown();
+    }
+    if (_multiStageWorkerRequestHandler != null) {
+      _multiStageWorkerRequestHandler.shutDown();
+    }
+  }
+
+  @Override
+  public BrokerResponse handleRequest(JsonNode request, @Nullable RequesterIdentity requesterIdentity,
```

Comment:
Do we not plan to implement the fallback / query failure based approach ?

So a JOIN query sent down the existing Netty based request handler will fail (hopefully with some deterministic error message). We catch that query and send it tt multistage engine if enabled ?

Reply:
no we are not implementing fallback as we parsed out the user-specified options and only use multi-stage when asked. so we no longer need fallback